### ISSUE TITLE
Use typehints

### ISSUE THREAD

**Issue by judgej** [proj4php/proj4php#6]
Typehints make sure the right type of classes are passed into methods and constructors. catching errors early, closer to source.

**Comment by judgej** [comment]
This is ongoing. Typehints at the moment are on the concrete classes, and will probably be completed using the concrete classes across the package.

A future release could look at better use of interfaces and abstract classes for better flexibility when used in projects.

**Comment by judgej** [comment]
Or maybe start using interfaces immediately? Any thoughts?

**Comment by julien2512** [comment]
Interface, Abstract, then Concrete is the good pattern, if it do not break existing code too much.

I guess I need to make a little more releases along your work !
Is that ok ?

**Comment by judgej** [comment]
Don't worry about trying to keep master in sync with the branch - they are all visible, the branches are still somewhat experimental (not all ideas will be good, so there could be some back-tracking). People can pull on any branch they want to try out or contribute to.

**Comment by julien2512** [comment]
Yes, and branches is designed for that : it leaves us a way to pull one or another.

So as to say, when your code is workable, and we are all agree, there is no reason not to pull it to master !

2.0.0 is done on #2 and #5.
